### PR TITLE
Cache dependencies for scala 2.13 [skip ci]

### PR DESCRIPTION
**Pull request by YanxuanLiu**
Fix https://github.com/NVIDIA/spark-rapids/issues/11199

1. changed `set-scala213-versions` to `cache-dependencies-scala213`, and added caching steps to this stage.
2. add `.github/workflows/mvn-verify-check/populate-daily-cache.sh` for caching dependencies for both scala 2.12 and 2.13.

### CODE REVIEW

**Comment by YanxuanLiu** [comment]
test build: https://github.com/YanxuanLiu/spark-rapids/actions/runs/10003551543

File: .github/workflows/mvn-verify-check/populate-daily-cache.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -x
+max_retry=3; delay=30; i=1
+while true; do
+for pom in pom.xml
```
Comment (pxLi) [inline .github/workflows/mvn-verify-check/populate-daily-cache.sh:20]: nit: bash format indentation

File: .github/workflows/mvn-verify-check/populate-daily-cache.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+set -x
+max_retry=3; delay=30; i=1
+while true; do
+for pom in pom.xml 
+do
+    mvn $COMMON_MVN_FLAGS --file $pom help:evaluate -pl dist \
+    -Dexpression=included_buildvers \
+    -DforceStdout -PnoSnapshots -q | tr -d ',' | \
+        xargs -n 1 bash -c \
+        "mvn $COMMON_MVN_FLAGS --file $1 -Dbuildver=$2 de.qaware.maven:go-offline-maven-plugin:resolve-dependencies" _ $pom
```
Comment (pxLi) [inline .github/workflows/mvn-verify-check/populate-daily-cache.sh:26]: https://github.com/NVIDIA/spark-rapids/actions/runs/10003888812/job/27651559232?pr=11231

```
https://github.com/NVIDIA/spark-rapids/actions/runs/10003888812/job/27651559232?pr=11231
```

File: .github/workflows/mvn-verify-check.yml [hunk 3]
```diff
@@ -167,16 +146,39 @@ jobs:
...
+        id: cache
+        uses: actions/cache@v4
+        with:
+          path: ~/.m2
+          key: ${{ env.scala213dailyCacheKey }}
+          restore-keys: ${{ runner.os }}-maven-
+      - name: populate-daily-cache
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: |
+          . .github/workflows/mvn-verify-check/populate-daily-cache.sh
```
Comment (pxLi) [inline .github/workflows/mvn-verify-check.yml:179]: go to `scala2.13` for the pom file

File: .github/workflows/mvn-verify-check/populate-daily-cache.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# limitations under the License.
+
+set -x
+max_retry=3; delay=30; i=1
+while true; do
+for pom in pom.xml 
+do
+    mvn $COMMON_MVN_FLAGS --file $pom help:evaluate -pl dist \
+    -Dexpression=included_buildvers \
+    -DforceStdout -PnoSnapshots -q | tr -d ',' | \
```
Comment (pxLi) [inline .github/workflows/mvn-verify-check/populate-daily-cache.sh:24]: `-PnoSnapshots` is for scala213, be sure about the shims list you are referring for different scala versions

**Review by pxLi — CHANGES_REQUESTED** [review]
There are many issues, please focus more on our objectives of setup the cache steps rather than merely passing the checks. Thank you.

**Comment by YanxuanLiu** [comment]
> There are many issues, please focus more on our objectives of setup the cache steps rather than merely passing the checks. Thank you.

Fixed the issues of parameters from previous PR https://github.com/NVIDIA/spark-rapids/pull/10994.

I set a env var `SCALA_VER` in github workflow, and the shell script will use appropriate pom file and `buildvers`.

File: .github/workflows/mvn-verify-check/populate-daily-cache.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+# limitations under the License.
+
+set -x
+max_retry=3; delay=30; i=1
+if [[ $SCALA_VER == '2.12' ]]; then
+    pom='pom.xml'
+    buildver='noSnapshots'
+elif [[ $SCALA_VER == '2.13' ]]; then
+    pom='scala2.13/pom.xml'
+    buildver='noSnapshotsScala213'
```
Comment (pxLi) [inline .github/workflows/mvn-verify-check/populate-daily-cache.sh:24]: nit: `noSnapshots` should auto-switch while different scala_version, so we may not need to retarget here. 

https://github.com/NVIDIA/spark-rapids/blob/branch-24.08/dist/pom.xml#L61-L70
https://github.com/NVIDIA/spark-rapids/blob/branch-24.08/scala2.13/dist/pom.xml#L61-L70

please help double confirm, thanks

**Review by pxLi — APPROVED** [review]
LGTM. Lets keep monitoring for a while after merge thanks

File: .github/workflows/mvn-verify-check.yml [hunk 6]
```diff
@@ -167,16 +148,41 @@ jobs:
             fi
           }
           done
-
-  set-scala213-versions:
+  
+
```
Comment (gerashegalov) [inline .github/workflows/mvn-verify-check.yml:152]: nit: extra newline
```suggestion
 
```

File: .github/workflows/mvn-verify-check.yml [hunk 7]
```diff
@@ -167,16 +148,41 @@ jobs:
...
-      - id: all213ShimVersionsStep
+      - uses: actions/setup-java@v4
+        with:
+          distribution: 'temurin'
+          java-version: 17
+      - name: Generate daily cache key
+        id: generateCacheKey
+        run: |
+          set -x
+          cacheKey="${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}-${{ github.event.pull_request.base.ref }}-$(date +'%Y-%m-%d')"
```
Comment (gerashegalov) [inline .github/workflows/mvn-verify-check.yml:169]: This is the same cacheKey as in the 2.12 job which could inadvertently result in a false cache hit. We could resolve it by either making sure that hashFiles receive only 2.12 or 2.13 poms exclusively when generating the corresponding key. Or we could simply inject 213 here in the key.

```suggestion
          cacheKey="${{ runner.os }}-maven-scala213-${{ hashFiles('**/pom.xml') }}-${{ github.event.pull_request.base.ref }}-$(date +'%Y-%m-%d')"
```

**Review by gerashegalov — DISMISSED** [review]

**Review by pxLi — APPROVED** [review]

File: .github/workflows/mvn-verify-check.yml [hunk 8]
```diff
@@ -167,16 +148,39 @@ jobs:
             fi
           }
           done
-
-  set-scala213-versions:
```
Comment (pxLi) [inline .github/workflows/mvn-verify-check.yml:171]: github commit suggestion got triggered twice and deleted the an extra line
https://github.com/NVIDIA/spark-rapids/pull/11231/commits/b19a8f2e17801dbc7fd8bd2bceb4941d23d6a5be

Comment (YanxuanLiu) [reply .github/workflows/mvn-verify-check.yml:171]: reverted the line

**Review by pxLi — APPROVED** [review]

**Pull request merged**